keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,211 @@
+For a highly available control plane to migrate from running only the KCM to running both the KCM and
+the CCM requires that only one process in the control plane is running the cloud specific controllers.
+This becomes non-trivial when introducing the CCM that runs overlapping controllers with the KCM.
+
+For environments that can tolerate downtime, the control plane can be taken down in order to
+reconfigure components to use the CCM, and then bring the control plane back up. This ensures that only
+1 component can be running the set of cloud controllers. For environments that have stricter requirements
+for uptime, some level of coordination is required between the two components to ensure that upgrading
+control planes does not result in running the same controller in more than one place while also accounting for version skew.
+
+In order to coordinate the cloud-specific controllers across the KCM and the CCM, this KEP proposes a
+_primary_ and N configurable _secondary_ (a.k.a migration) leader election locks in the KCM and the CCM.
+The primary lock represents the current leader election resource lock in the KCM and the CCM. The set of
+secondary locks are defined by the cloud provider and run in parallel to the primary locks. For a migration
+lock defined by the cloud provider, the cloud provider also determines the set of controllers run within the
+migration lock and the controller manager it should run in - either the CCM or the KCM.
+
+The properties of the migration lock are:
+  * must have a unique name
+  * the set of controllers in the lock is immutable.
+  * no two migration locks should have overlapping controllers
+  * the controller manager where the lock runs can change across releases.
+  * can only run in either the kube-controller-manager or the cloud-controller-manager.
+
+During migration, either the KCM or CCM may have multiple migration locks, though for performance reasons no more than 3 locks is recommended.
+
+Let's say we are migrating the service, route, and nodeipam controllers from the KCM to the CCM across Kubernetes versions, say v1.17 to v1.18.
+In v1.17, the cloud provider would define a new migration lock called `cloud-network-controller-migration` which specifies those controllers to run
+inside the KCM (see Figure 1). As a result, in v1.17 those controllers would run in the KCM but under the `cloud-network-controller-migration` leader election.
+To migrate to the CCM for v1.18, the cloud provider would update the `cloud-network-controller-migration` lock to now run in the CCM (see Figure 2). During migration,
+only one instance of `cloud-network-controller-migration` can run and once the control plane is fully migrated to v1.18, it should only run
+in the CCM as specified by the cloud provider. During a control plane upgrade, the cloud network controllers may still run in one of the KCMs that are still
+on v1.17, but none of the v1.18 CCMs can claim the lock.
```

Comment:
I agree but I think the sentence is a little confusing.

During a control plane upgrade, the cloud network controllers may still run in one of the KCMs that are still
on v1.17. A 1.17 KCM holding the lock will prevent any of the v1.18 CCMs from claiming the lock. When the current holder of the lock goes down, one of the controller managers eligible will acquire lock.

Reply:
That's much better, udpated!